providers/implementations/keymgmt/rsa_kmgmt.c

```diff
@@ -239,7 +244,96 @@ static const OSSL_PARAM *rsa_importkey_types(void)
     return rsa_key_types;
 }
 
+struct rsa_genkey_ctx {
+    size_t nbits;
+    BIGNUM *pub_exp;
+    size_t primes;
+};
+
+static int rsa_gencb(int p, int n, BN_GENCB *cb)
+{
+    OSSL_CALLBACK *arg = BN_GENCB_get_arg(cb);
+    OSSL_PARAM params[] = { OSSL_PARAM_END, OSSL_PARAM_END, OSSL_PARAM_END };
+
+    params[0] = OSSL_PARAM_construct_int(OSSL_GEN_PARAM_POTENTIAL, &p);
+    params[1] = OSSL_PARAM_construct_int(OSSL_GEN_PARAM_ITERATION, &n);
+
+    return c_generic_callback(arg, params);
+}
+
+static void *rsa_genkey_init(void *provctx)
+{
+    struct rsa_genkey_ctx *gctx = OPENSSL_zalloc(sizeof(*gctx));
+
+    if (gctx != NULL) {
+        if ((gctx->pub_exp = BN_new()) == NULL
+            || !BN_set_word(gctx->pub_exp, RSA_F4)) {
+            BN_free(gctx->pub_exp);
+            gctx = NULL;
+        } else {
+            gctx->nbits = 2048;
+            gctx->primes = RSA_DEFAULT_PRIME_NUM;
+        }
+    }
+    return gctx;
+}
+
+static int rsa_genkey_set_params(void *genctx, const OSSL_PARAM params[])
+{
+    struct rsa_genkey_ctx *gctx = genctx;
+    const OSSL_PARAM *p;
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_BITS)) != NULL
+        && !OSSL_PARAM_get_size_t(p, &gctx->nbits))
+        return 0;
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_PRIMES)) != NULL
+        && !OSSL_PARAM_get_size_t(p, &gctx->primes))
+        return 0;
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_E)) != NULL
+        && !OSSL_PARAM_get_BN(p, &gctx->pub_exp))
+        return 0;
+    return 1;
+}
+
+static void *rsa_genkey(void *genctx, void *domparams, OSSL_CALLBACK *osslcb)
+{
+    struct rsa_genkey_ctx *gctx = genctx;
+    RSA *rsa = RSA_new();
+    BN_GENCB *gencb = NULL;
+
+    gencb = BN_GENCB_new();
+    if (gencb != NULL)
+        BN_GENCB_set(gencb, rsa_gencb, osslcb);
+
+    if (gctx == NULL)
+        return NULL;
+
+    if (!RSA_generate_multi_prime_key(rsa, (int)gctx->nbits, (int)gctx->primes,
```

Comment:
We need to make sure that multi prime keys are not used in fips mode.. You should see FIPS specific code in the existing rsa keygen code.

Reply:
Okie